src/frontend/src/optimizer/plan_node/generic/share.rs

```diff
@@ -12,52 +12,61 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::cell::RefCell;
 use std::hash::Hash;
+use std::sync::Arc;
 
+use parking_lot::RwLock;
 use risingwave_common::catalog::Schema;
 
 use super::{GenericPlanNode, GenericPlanRef};
 use crate::OptimizerContextRef;
 use crate::optimizer::property::FunctionalDependencySet;
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone)]
 pub struct Share<PlanRef> {
-    pub input: RefCell<PlanRef>,
+    pub input: Arc<RwLock<PlanRef>>,
 }
 
+impl<P: PartialEq> PartialEq for Share<P> {
+    fn eq(&self, other: &Self) -> bool {
+        *self.input.read() == *other.input.read()
```

Comment:
Here might be the cause of the deadlock mentioned in our offline discussion, when the two input refers to the same lock.